linera-core/src/client.rs

```diff
@@ -1294,17 +1316,215 @@ where
+    }
+
+    pub async fn try_synchronize_chain_state_from<A>(
+        &self,
+        name: ValidatorName,
+        mut node: A,
+        chain_id: ChainId,
+        notifications: &mut impl Extend<Notification>,
+    ) -> Result<(), ChainClientError>
+    where
+        A: LocalValidatorNode + Clone + 'static,
+    {
+        let local_info = self.client.local_node.local_chain_info(chain_id).await?;
+        let range = BlockHeightRange {
+            start: local_info.next_block_height,
+            limit: None,
+        };
+        let query = ChainInfoQuery::new(chain_id)
+            .with_sent_certificate_hashes_in_range(range)
+            .with_manager_values();
+        let info = match node.handle_chain_info_query(query).await {
+            Ok(response) if response.check(name).is_ok() => response.info,
+            Ok(_) => {
+                tracing::warn!("Ignoring invalid response from validator");
+                // Give up on this validator.
+                return Ok(());
+            }
+            Err(err) => {
+                tracing::warn!("Ignoring error from validator: {}", err);
+                return Ok(());
+            }
+        };
+
+        let certificates = future::try_join_all(
+            info.requested_sent_certificate_hashes
+                .into_iter()
+                .map(|hash| {
+                    let mut node = node.clone();
+                    async move { node.download_certificate(hash).await }
+                }),
+        )
+        .await?;
+
+        if !certificates.is_empty()
+            && self
+                .client
+                .local_node
+                .try_process_certificates(name, &mut node, chain_id, certificates, notifications)
+                .await
+                .is_none()
+        {
+            return Ok(());
+        };
+        if let Some(proposal) = info.manager.requested_proposed {
+            if proposal.content.block.chain_id == chain_id {
+                let owner = proposal.owner;
+                loop {
+                    let result = self
+                        .client
+                        .local_node
+                        .handle_block_proposal(*proposal.clone())
+                        .await;
+                    if let Err(original_err) = &result {
+                        if let LocalNodeError::WorkerError(WorkerError::ChainError(chain_error)) =
+                            original_err
+                        {
+                            if let ChainError::ExecutionError(
+                                ExecutionError::SystemError(
+                                    SystemExecutionError::BlobNotFoundOnRead(blob_id),
+                                ),
+                                _,
+                            ) = &**chain_error
+                            {
+                                self.update_local_node_with_blob(*blob_id).await?;
+                                continue;
```

Comment:
I think we should also handle the case where a _published_ blob is not found, and download those from the validators. Can be in a different PR, though.

Reply:
But what if the client downloads a block _proposal_ from a validator (the `proposed` entry in the manager) and the client doesn't have a blob in that block? Then it should download that blob from that validator.

Reply:
Isn't that what's happening in this code? 🤔

Reply:
I don't think so: This code is for when a _read_ fails, and then (as it should) finds a confirmed certificate that used that blob, processes it, and thus also downloads the blob.

But what if the proposal tries to publish a new block, and we don't have it?
So: Another client proposed a block that would publish a blob, but somehow failed (maybe didn't reach enough validators), and now we got that proposal from a validator and try to add it to our local node, so we can later re-propose it, and so that our local chain manager knows that we shouldn't propose something else in this round. Our local node refuses to add it because we don't have the blob. So we need to download the blob from that validator (without any certificate, because this is a new blob about to be published).

Reply:
I'm not sure I get it. Do you mean another owner in the same chain? If so then the local node will already have it. If an owner in a chain tries to publish a blob, fails, then it gets retried by a different chain, then AFAIU it won't work (and I don't think it's supposed to work?) and we also don't test that case.